ludwig/utils/torch_utils.py

```diff
@@ -93,6 +93,23 @@ def losses(self):
             collected_losses.extend(layer.losses)
         return collected_losses
 
+
+class LudwigComponent(Module):
+    def __init__(self):
+        super().__init__()
+
+    @property
+    def input_shape(self) -> torch.Size:
+        """ Returns the size of the input tensor without the batch dimension. """
+        raise NotImplementedError('Abstract class.')
+
+    @property
+    def output_shape(self) -> torch.Size:
+        """ Returns the size of the output tensor without the batch dimension."""
+        output_tensor = self.forward(torch.rand(*self.input_shape))
```

Comment:
General advice is to avoid calling `forward` directly. Not sure it applies here, but probably good practice to follow:

https://discuss.pytorch.org/t/any-different-between-model-input-and-model-forward-input/3690

Reply:
I think the advantage of using `__call__` instead of `forward` is to call the forward and backward hooks. In our case it's probably more efficient to not call the hooks since we only want to use forward to compute the shapes, if in case the forward and backward hooks end up having some expensive computation.

I'm happy to switch to the `__call__` method if you prefer though.

Reply:
Since in most cases the hooks are just side effects, we probably don't want to call them here, I agree. So for now I think we can stick with forward.